File: Java/src/main/java/org/example/RomanNumeral.java
package org.example;

import java.util.List;
import java.util.Objects;
import java.util.function.BinaryOperator;

public class RomanNumeral {

    private final int value;

    public RomanNumeral(int value) {
        if (value < 1 || value > 3999) {
            throw new IllegalArgumentException(
                    String.format("%d is out of range for roman numerals.", value));
        }
        this.value = value;
    }

    public RomanNumeral(String roman) {
        value = parse(roman);
    }

    public RomanNumeral plus(RomanNumeral toAdd) {
        return new RomanNumeral(this.value + toAdd.value);
    }

    public int toInt() {
        return value;
    }

    @Override
    public String toString() {
        return convertToRoman(value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        RomanNumeral that = (RomanNumeral) o;
        return value == that.value;
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    private int parse(String roman) {
        final var thousandsPart = highestMatchingPrefixOrNull(roman, allRomanThousands);
        final var thousandsValue = toThousandsValueOrZero(thousandsPart);
        final var remainingRomanWithoutThousands = roman.replaceFirst(
                thousandsPart != null ? thousandsPart : "",
                ""
        );

        final var hundredsPart = highestMatchingPrefixOrNull(remainingRomanWithoutThousands, allRomanHundreds);
        final var hundredsValue = toHundredsValueOrZero(hundredsPart);
        final var remainingRomanWithoutHundreds = remainingRomanWithoutThousands.replaceFirst(
                hundredsPart != null ? hundredsPart : "",
                ""
        );

        final var tensPart = highestMatchingPrefixOrNull(remainingRomanWithoutHundreds, allRomanTens);
        final var tensValue = toTensValueOrZero(tensPart);
        final var remainingRomanWithoutTens = remainingRomanWithoutHundreds.replaceFirst(
                tensPart != null ? tensPart : "",
                ""
        );

        final var unitsPart = highestMatchingPrefixOrNull(remainingRomanWithoutTens, allRomanUnits);
        final var unitsValue = toUnitsValueOrZero(unitsPart);
        final var remainingRomanWithoutUnits = remainingRomanWithoutTens.replaceFirst(
                unitsPart != null ? unitsPart : "",
                ""
        );

        final var parsedValue = thousandsValue + hundredsValue + tensValue + unitsValue;

        if (parsedValue == 0 || !remainingRomanWithoutUnits.isEmpty()) {
            throw new NumberFormatException(
                    String.format("%s is not a supported roman numeral.", roman));
        }

        return parsedValue;
    }

    private String highestMatchingPrefixOrNull(String roman, List<String> positionalPrefixes) {
        // We match the last, i.e. the highest prefix for the current decimal position,
        // since higher roman "digits" can contain lower ones (e. g. XC contains X),
        // which would otherwise lead to a false-positive matching of X.
        BinaryOperator<String> reduceToLast = (first, second) -> second;

        return positionalPrefixes
                .stream()
                .filter(roman::startsWith)
                .reduce(reduceToLast)
                .orElse(null);
    }

    private int toThousandsValueOrZero(String romanThousandsPartOrNull) {
        if (romanThousandsPartOrNull == null) {
            return 0;
        }

        final var index = allRomanThousands.indexOf(romanThousandsPartOrNull);
        if (index < 0) {
            throw new NumberFormatException();
        }

        return 1000 * (index + 1);
    }

    private int toHundredsValueOrZero(String romanHundredsPartOrNull) {
        if (romanHundredsPartOrNull == null) {
            return 0;
        }

        final var index = allRomanHundreds.indexOf(romanHundredsPartOrNull);
        if (index < 0) {
            throw new NumberFormatException();
        }

        return 100 * (index + 1);
    }

    private int toTensValueOrZero(String romanTensPartOrNull) {
        if (romanTensPartOrNull == null) {
            return 0;
        }

        final var index = allRomanTens.indexOf(romanTensPartOrNull);
        if (index < 0) {
            throw new NumberFormatException();
        }

        return 10 * (index + 1);
    }

    private int toUnitsValueOrZero(String romanUnitsPartOrNull) {
        if (romanUnitsPartOrNull == null) {
            return 0;
        }

        final var index = allRomanUnits.indexOf(romanUnitsPartOrNull);
        if (index < 0) {
            throw new NumberFormatException();
        }

        return index + 1;
    }

    private String convertToRoman(int n) {
        // Offset for zero-based indices, e.g. "I" is at position 0.
        final var indexOffset = 1;

        final var thousandsIndex = n / 1000 - indexOffset;
        final var romanThousands = getOrElse(allRomanThousands, thousandsIndex, "");

        final var hundredsIndex = (n % 1000) / 100 - indexOffset;
        final var romanHundreds = getOrElse(allRomanHundreds, hundredsIndex, "");

        final var tensIndex = (n % 100) / 10 - indexOffset;
        final var romanTens = getOrElse(allRomanTens, tensIndex, "");

        final var unitsIndex = n % 10 - indexOffset;
        final var romanUnits = getOrElse(allRomanUnits, unitsIndex, "");

        return romanThousands + romanHundreds + romanTens + romanUnits;
    }

    private <T> T getOrElse(List<T> list, int index, T defaultValue) {
        if (index > list.size() - 1) {
            throw new IndexOutOfBoundsException(index);
        } else if (index < 0) {
            return defaultValue;
        } else {
            return list.get(index);
        }
    }

    private final List<String> allRomanThousands = List.of(
            "M",
            "MM",
            "MMM"
    );

    private final List<String> allRomanHundreds = List.of(
            "C",
            "CC",
            "CCC",
            "CD",
            "D",
            "DC",
            "DCC",
            "DCCC",
            "CM"
    );

    private final List<String> allRomanTens = List.of(
            "X",
            "XX",
            "XXX",
            "XL",
            "L",
            "LX",
            "LXX",
            "LXXX",
            "XC"
    );

    private final List<String> allRomanUnits = List.of(
            "I",
            "II",
            "III",
            "IV",
            "V",
            "VI",
            "VII",
            "VIII",
            "IX"
    );
}

File: Kotlin/src/main/kotlin/RomanNumeral.kt
package org.example

class RomanNumeral {

    private val value: Int

    constructor(value: Int) {
        require(value in 1..3999) {
            String.format("%d is out of range for roman numerals.", value)
        }
        this.value = value
    }

    constructor(roman: String) {
        value = parse(roman)
    }

    operator fun plus(toAdd: RomanNumeral): RomanNumeral {
        return RomanNumeral(value + toAdd.value)
    }

    fun toInt(): Int {
        return value;
    }

    override fun toString(): String {
        return convertToRoman(value)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as RomanNumeral
        if (value != other.value) return false
        return true
    }

    override fun hashCode(): Int = value

    private fun parse(roman: String): Int {
        val thousandsPart = highestMatchingPrefix(roman, allRomanThousands)
        val thousandsValue = toThousandsValueOrNull(thousandsPart) ?: 0
        val remainingRomanWithoutThousands = roman.removePrefix(thousandsPart ?: "")

        val hundredsPart = highestMatchingPrefix(remainingRomanWithoutThousands, allRomanHundreds)
        val hundredsValue = toHundredsValueOrNull(hundredsPart) ?: 0
        val remainingRomanWithoutHundreds = remainingRomanWithoutThousands.removePrefix(hundredsPart ?: "")

        val tensPart = highestMatchingPrefix(remainingRomanWithoutHundreds, allRomanTens)
        val tensValue = toTensValueOrNull(tensPart) ?: 0
        val remainingRomanWithoutTens = remainingRomanWithoutHundreds.removePrefix(tensPart ?: "")

        val unitsPart = highestMatchingPrefix(remainingRomanWithoutTens, allRomanUnits)
        val unitsValue = toUnitsValueOrNull(unitsPart) ?: 0
        val remainingRomanWithoutUnits = remainingRomanWithoutTens.removePrefix(unitsPart ?: "")

        val parsedValue = thousandsValue + hundredsValue + tensValue + unitsValue

        if (parsedValue == 0 || remainingRomanWithoutUnits.isNotBlank()) {
            throw NumberFormatException("`$roman` is not a valid roman numeral.")
        }

        return parsedValue
    }

    private fun highestMatchingPrefix(
        roman: String,
        positionalPrefixes: List<String>
    ): String? {
        // We match the last, i. e. highest prefix for the current decimal position,
        // since higher roman "digits" can contain lower ones (e. g. XC contains X),
        // which would otherwise lead to a false-positive matching of X.
        return positionalPrefixes.lastOrNull { roman.startsWith(it) }
    }

    private fun toThousandsValueOrNull(romanThousandsPartOrNull: String?): Int? {
        romanThousandsPartOrNull ?: return null
        val index = allRomanThousands.indexOf(romanThousandsPartOrNull)
        if (index < 0) {
            throw NumberFormatException()
        }
        return 1000 * (index + 1)
    }

    private fun toHundredsValueOrNull(romanHundredsPartOrNull: String?): Int? {
        romanHundredsPartOrNull ?: return null
        val index = allRomanHundreds.indexOf(romanHundredsPartOrNull)
        if (index < 0) {
            throw NumberFormatException()
        }
        return 100 * (index + 1)
    }

    private fun toTensValueOrNull(romanTensPartOrNull: String?): Int? {
        romanTensPartOrNull ?: return null
        val index = allRomanTens.indexOf(romanTensPartOrNull)
        if (index < 0) {
            throw NumberFormatException()
        }
        return 10 * (index + 1)
    }

    private fun toUnitsValueOrNull(romanUnitsPartOrNull: String?): Int? {
        romanUnitsPartOrNull ?: return null
        val index = allRomanUnits.indexOf(romanUnitsPartOrNull)
        if (index < 0) {
            throw NumberFormatException()
        }
        return 1 * (index + 1)
    }

    private fun convertToRoman(n: Int): String {
        // Offset for zero-based indices, e.g. "I" is at position 0.
        val indexOffset = 1

        val thousandsIndex = n / 1000 - indexOffset
        val romanThousands = allRomanThousands.getOrElse(thousandsIndex) { "" }

        val hundredsIndex = (n % 1000) / 100 - indexOffset
        val romanHundreds = allRomanHundreds.getOrElse(hundredsIndex) { "" }

        val tensIndex = (n % 100) / 10 - indexOffset
        val romanTens = allRomanTens.getOrElse(tensIndex) { "" }

        val unitsIndex = n % 10 - 1
        val romanUnits = allRomanUnits.getOrElse(unitsIndex) { "" }

        return "$romanThousands$romanHundreds$romanTens$romanUnits"
    }

    private val allRomanThousands: List<String> =
        listOf(
            "M",
            "MM",
            "MMM"
        )

    private val allRomanHundreds: List<String> =
        listOf(
            "C",
            "CC",
            "CCC",
            "CD",
            "D",
            "DC",
            "DCC",
            "DCCC",
            "CM"
        )

    private val allRomanTens: List<String> =
        listOf(
            "X",
            "XX",
            "XXX",
            "XL",
            "L",
            "LX",
            "LXX",
            "LXXX",
            "XC"
        )

    private val allRomanUnits: List<String> =
        listOf(
            "I",
            "II",
            "III",
            "IV",
            "V",
            "VI",
            "VII",
            "VIII",
            "IX"
        )
}
